### ISSUE TITLE
Unexpected exception thrown in parsing run after Compact & Repair

### ISSUE THREAD

**Issue by alexkadis** [rubberduck-vba/Rubberduck#3787]
When opening Access for the first time #2915 is resolved. But if I run Compact & Repair, a similar error occurs:

```
2018-02-22 15:28:18.9582;ERROR-2.1.2.2870;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 14).;System.Runtime.InteropServices.COMException (0x80010105): The server threw an exception. (Exception from HRESULT: 0x80010105 (RPC_E_SERVERFAULT))
   at Microsoft.Vbe.Interop.VBProjectClass.get_Protection()
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerProjectViewModel..ctor(FolderHelper folderHelper, Declaration declaration, IEnumerable`1 declarations) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerProjectViewModel.cs:line 37
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.<HandleStateChanged>b__51_2(IGrouping`2 grouping) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 319
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.HandleStateChanged(Object sender, ParserStateEventArgs e) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 318
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 244
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 378
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 313
```

Opening as a new issue at the request of @bclothier

**Comment by Vogel612** [comment]
Related: #3779 for better handling of errors occuring where this occurs

**Comment by bclothier** [comment]
I can reproduce this on Access 2010.

```Exception thrown at 0x0F8C2149 (VBE7.DLL) in MSACCESS.EXE: 0xC0000005: Access violation reading location 0x00000210.```

```System.AccessViolationException
  HResult=0x80004003
  Message=Attempted to read or write protected memory. This is often an indication that other memory is corrupt.
  Source=<Cannot evaluate the exception source>
  StackTrace:
<Cannot evaluate the exception stack trace>
```

![image](https://user-images.githubusercontent.com/2367644/36580245-1217659c-182d-11e8-8940-5bf3a6f6bad7.png)


It appears to me the blame lies with the test explorer trying to evaluate whether it can refresh parse, presumably after the database has become unavailable during the Compact & Repair.

**Comment by WaynePhillipsEA** [comment]
does it not fire ProjectRemoved event before starting the C&R?

**Comment by bclothier** [comment]
I checked and it is not even fired. Perhaps it does get eventually, but the access violation occurs before the breakpoint is hit. This may be a similar symptom to what @MDoerner  was observing with Excel; strong references to a COM object (perhaps the `VBProject` or one of its child) may prevent it being released and thus the event won't fire.

**Comment by bclothier** [comment]
Oh, BTW, I also verified that if I load Rubberduck and do *NOT* parse, the C & R executes without a problem.

However, when it reloads, Rubberduck automatically parses the project being re-loaded, so if I then C & R 2nd time it will have that access violation.

Furthermore, I was wrong to accuse the test explorer because in this crash I'm looking at, it's also Todo Explorer. The common culprit is the `RubberduckParserState.IsDirty` method, which gets called as part of the `RefreshCommand` so it's not a particular tool window, but rather that WPF is calling `CanExecute` as an effect of having executing C & R, and the `IsDirty` attempts to enumerate the projects & components, which it can't.

**Comment by Vogel612** [comment]
soo ... the fix is to get `IsDirty` to not try to enumerate projects & components which it can't enumerate?

**Comment by bclothier** [comment]
I think my first testing was flawed. The `ProjectAdded` and `ProjectRemoved` *do* get fired. In fact they are fired 3 times each, (removed, added, removed, added, removed, added) 

I also verified that the `ProjectRepository` also got its `Refresh` called. 

In spite of all above, AV occurs.

**Comment by WaynePhillipsEA** [comment]
The main issue I can see is that ParseCoordinator::ParseAllInternal takes a snapshot of the VBProjects etc in its worker thread (in the projectsRepository).

Meanwhile, on the main thread we can receive a ProjectRemoved event for a project whilst the other threads are still running and issuing calls to the said project COM object.

This wouldn't ordinarily be a problem if the VBE didn't destroy objects until their ref counts reached zero, but alas this is the VBE, and it does what it wants.

For this reason, at the point we receive the ProjectRemoved event, and before returning from it, we must wait for the parse threads to finish (or confirm they are cancelled and hands off the COM objects).

Only once we are sure we wont issue any more calls to the COM object that is being removed, can we can safely return from the ProjectRemoved event.

**Comment by WaynePhillipsEA** [comment]
BTW, as is typically with these types of unmanaged/COM bugs, I couldn't repro exactly.  Mine completes the C&R, but starts crashing at the auto-parse.

**Comment by bclothier** [comment]
It bears noting that a prior version of RD had auto-parsing on load disabled because we cannot reliably determine when it's safe to parse. At some point, it got enabled again.

**Comment by WaynePhillipsEA** [comment]
From what I understand, a lot of trial & error has happened over the years in RD.  But a lot has changed recently, so what was thought of as being a VBE bug might not have been, I guess.

**Comment by WaynePhillipsEA** [comment]
To get this right looks to be quite involved, at least from what I can see.   To synchronously, and thoroughly ensure all project, component, code modules etc are released at the point we receive the ProjectRemoved event is going to take some effort.

It's certainly not just a case of waiting for the parser threads to end (as I alluded to earlier), as the ProjectRepository looks to be used extensively via the parser state.

Someone with better knowledge of the RD codebase might be able to suggest something.

Arguably, we should also be doing the same for the ComponentRemoved event (i.e. ensuring that all COM references of that component are released before returning from that event)...

**Comment by WaynePhillipsEA** [comment]
So an update:  the reason this is occuring now is due to the recent changes in QMNs, with them no 
longer holding on to references to components of the project.

After returning from the ProjectRemoved event, the VBE starts internal destruction of the project as much as it can.  As
there are no longer any open references to the Components collection, nor its elements, it decides to destroy that 
collection, presuming that no-one will ask for it now that it has sent out the ProjectRemoved event.  

Previously, the QMNs were keeping the VBComponents alive, blocking the VBE from destroying them at this point.  

When we later query the VBComponents through the cached ProjectsRepository in other threads (e.g. parsers), this 
can cause a hard crash, as the VBE has destroyed the underlying collection without adding safeguards to the getter.

We can confirm that keeping the VBComponents open is enough to prevent the crash (at least on my machine) by adding 
this to the ProjectRemoved event:
    Marshal.GetIUnknownForObject(e.Project.VBComponents.Target);

^ this is not a suitable fix.

Being a Good COM Citizen™ is starting to bite us on the bum.

**Comment by alexkadis** [comment]
Out of curiosity, could Rubberduck keep track of when a Compact & Repair happens and then completely reload itself from scratch?

**Comment by bclothier** [comment]
Well, the issue is that Rubberduck is a *VBE* Add-in, not an Access Add-in (nor an Excel Add-in, and so on). In order for it to "know" about the compacting and repairing, it'd have to hook into Access' object model. We kind of can do that but that violates the notion of being just a VBE Add-In. We try to minimize host-specific stuff as much as possible. 

Could you do me a favor and try [the last green release 2.1](https://github.com/rubberduck-vba/Rubberduck/releases/tag/v2.1.1)?  and confirm if the symptom persist even with that release?

**Comment by alexkadis** [comment]
@bclothier ok just tried the stable 2.1.1 release and still having this issue. I'm also getting the parse error that was blocking me until  v2.1.2.2870.

```
2018-02-26 08:54:20.4394;ERROR-2.1.1.2532;Rubberduck.Common.LogLevelHelper;
	Rubberduck version 2.1.1.2532 loading:
	Operating System: Microsoft Windows NT 10.0.16299.0 x64
	Host Product: Microsoft Office 2013 x64
	Host Version: 15.0.4963.1000
	Host Executable: MSACCESS.EXE;
2018-02-26 08:55:10.4682;ERROR-2.1.1.2532;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 13).;System.Runtime.InteropServices.COMException (0x80010105): The server threw an exception. (Exception from HRESULT: 0x80010105 (RPC_E_SERVERFAULT))
   at Microsoft.Vbe.Interop.VBProjectClass.get_Protection()
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerProjectViewModel..ctor(FolderHelper folderHelper, Declaration declaration, IEnumerable`1 declarations) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerProjectViewModel.cs:line 39
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.<HandleStateChanged>b__52_3(IGrouping`2 grouping) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 326
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.HandleStateChanged(Object sender, ParserStateEventArgs e) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 325
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 244
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 378
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 313
2018-02-26 08:55:19.7889;ERROR-2.1.1.2532;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 27).;System.Runtime.InteropServices.COMException (0x80010105): The server threw an exception. (Exception from HRESULT: 0x80010105 (RPC_E_SERVERFAULT))
   at Microsoft.Vbe.Interop.VBProjectClass.get_Protection()
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerProjectViewModel..ctor(FolderHelper folderHelper, Declaration declaration, IEnumerable`1 declarations) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerProjectViewModel.cs:line 39
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.<HandleStateChanged>b__52_3(IGrouping`2 grouping) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 326
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.HandleStateChanged(Object sender, ParserStateEventArgs e) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 325
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 244
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 378
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 313
2018-02-26 08:57:04.9683;ERROR-2.1.1.2532;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 15).;System.Runtime.InteropServices.COMException (0x80010105): The server threw an exception. (Exception from HRESULT: 0x80010105 (RPC_E_SERVERFAULT))
   at Microsoft.Vbe.Interop.VBProjectClass.get_Protection()
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerProjectViewModel..ctor(FolderHelper folderHelper, Declaration declaration, IEnumerable`1 declarations) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerProjectViewModel.cs:line 39
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.<HandleStateChanged>b__52_3(IGrouping`2 grouping) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 326
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.HandleStateChanged(Object sender, ParserStateEventArgs e) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 325
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 244
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 378
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 313
```

**Comment by alexkadis** [comment]
Also tried with .2886 (latest pre-release) and still experiencing the error.

```
2018-02-26 09:00:50.3771;ERROR-2.1.2.2886;Rubberduck.Common.LogLevelHelper;
	Rubberduck version 2.1.2.2886 loading:
	Operating System: Microsoft Windows NT 10.0.16299.0 x64
	Host Product: Microsoft Office 2013 x64
	Host Version: 15.0.4963.1000
	Host Executable: MSACCESS.EXE;
2018-02-26 09:01:14.1852;ERROR-2.1.2.2886;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 10).;System.Runtime.InteropServices.COMException (0x80010105): The server threw an exception. (Exception from HRESULT: 0x80010105 (RPC_E_SERVERFAULT))
   at Microsoft.Vbe.Interop.VBProjectClass.get_Protection()
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerProjectViewModel..ctor(FolderHelper folderHelper, Declaration declaration, IEnumerable`1 declarations) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerProjectViewModel.cs:line 37
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.<HandleStateChanged>b__51_2(IGrouping`2 grouping) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 319
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.HandleStateChanged(Object sender, ParserStateEventArgs e) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 318
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 244
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 378
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 313
2018-02-26 09:01:17.0662;ERROR-2.1.2.2886;Rubberduck.Parsing.VBA.ParseCoordinator;Unexpected exception thrown in parsing run. (thread 24).;System.Runtime.InteropServices.COMException (0x80010105): The server threw an exception. (Exception from HRESULT: 0x80010105 (RPC_E_SERVERFAULT))
   at Microsoft.Vbe.Interop.VBProjectClass.get_Protection()
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerProjectViewModel..ctor(FolderHelper folderHelper, Declaration declaration, IEnumerable`1 declarations) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerProjectViewModel.cs:line 37
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.<HandleStateChanged>b__51_2(IGrouping`2 grouping) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 319
   at System.Linq.Enumerable.WhereSelectListIterator`2.MoveNext()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at Rubberduck.Navigation.CodeExplorer.CodeExplorerViewModel.HandleStateChanged(Object sender, ParserStateEventArgs e) in C:\projects\rubberduck\RetailCoder.VBE\Navigation\CodeExplorer\CodeExplorerViewModel.cs:line 318
   at System.EventHandler`1.Invoke(Object sender, TEventArgs e)
   at Rubberduck.Parsing.VBA.ParseCoordinator.ExecuteCommonParseActivities(IReadOnlyCollection`1 toParse, IReadOnlyCollection`1 toReresolveReferencesInput, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 244
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAllInternal(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 378
   at Rubberduck.Parsing.VBA.ParseCoordinator.ParseAll(Object requestor, CancellationToken token) in C:\projects\rubberduck\Rubberduck.Parsing\VBA\ParseCoordinator.cs:line 313
```

**Comment by bclothier** [comment]
Thanks for confirming. This is consistent with what we've been observing; I wanted an independent attestation that this wasn't something new introduced to recent versions because honestly, I was surprised that it wasn't working. That said, we will be working on a fix but it will be few days. I do apologize for the inconvenience.

**Comment by alexkadis** [comment]
@bclothier no need to apologize! This is an open-source free tool for me, if there's any small way I can contribute, please just ask.

**Comment by bclothier** [comment]
Just to clarify - the surprise was with inability to do a C & R after a parsing. The other error regarding getting properties off the `VBComponents`, that was indeed introduced by a recent change in order to avoid memory leaks.

**Comment by alexkadis** [comment]
Yup! Makes sense to me. I'm able to C&R, just can't re-parse _after_ a C&R (I can continue to work or C&R, it's that Rubberduck is frozen, not Access itself).

**Comment by MDoerner** [comment]
I just had a look at the code throwing the error. The problem is not the `ProjectsRepository`, which should get refreshed just fine. The problem are the project references saved on the declarations. These still point to the old object.

I guess we are not refreshing these because the project events fire in rapid succession such that we never get to the clean up part before cancellation of the parsing run. Then, after all events have fired, all projects are there again, with the same project ids. Thus, they do not get reparsed.

Maybe, we could just add all modules in inaccessible projects to the list of modules to reparse. That would at least make the error recoverable by parsing again.

**Comment by bclothier** [comment]
Cannot reproduce anymore. Closing this as solved. If anyone is still observing this, feel free to reopen.
